Use simplify_exprtt::resultt in pre-order simplification steps

tautschnig opened the pull request:
The use of resultt increases type safety as the expression to be
simplified is no longer modified in place. All post-order simplification
steps already use resultt, but pre-order steps had been left to be
done.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Use simplify_exprtt::resultt in pre-order simplification steps

The use of resultt increases type safety as the expression to be
simplified is no longer modified in place. All post-order simplification
steps already use resultt, but pre-order steps had been left to be
done.

kroening commented on src/util/simplify_expr_if.cpp:
```diff
@@ -260,21 +279,18 @@ bool simplify_exprt::simplify_if_preorder(if_exprt &expr)
+    auto r_truevalue = simplify_rec(swap_branches ? falsevalue : truevalue);
+    if(swap_branches)
+      r_truevalue.expr_changed = CHANGED;
 
```
Why is this needed? The if-expression will be marked as changed anyway.

tautschnig replied:
Yes, but `build_if_expr` would not actually replace the `truevalue`.

tautschnig replied:
I have refactored the code and added a comment.

kroening commented on src/util/simplify_expr_if.cpp:
```diff
@@ -283,48 +299,36 @@ bool simplify_exprt::simplify_if_preorder(if_exprt &expr)
+    auto falsevalue = simplify_rec(swap_branches ? falsevalue : truevalue);
+    if(swap_branches)
+      r_falsevalue.expr_changed = CHANGED;
 
```
Same here

tautschnig replied:
`build_if_expr` would not actually replace the `falsevalue`.

tautschnig committed:
Simplify selected ID_if operands in preorder traversal

This avoids repeatedly visiting already-simplified operands. On the
example from #7357 this reduces symex time from 1172 seconds to 922
seconds.
